doc/developer/design/20230306_persist_mfp_pushdown.md

```diff
@@ -0,0 +1,272 @@
+
+There are a number of options here, including machine learning models, but simple heuristics are understandable and can get us surprisingly far. It is difficult to predict exactly what the heuristics will be, but an initial thought is something like the following:
+
+- First, a budget (in bytes) is calculated, possibly as a function of how many updates are in the part. We guarantee that the resulting statistics fit into this budget.
+- Second, anything obviously unnecessary is trimmed. For example, any real query filtering on a json column is likely to have a cast in the expression tree. This means that many (all?) json stats computed on a column containing more than one Datum type (e.g. both `Numeric` and `String`) would output an error from MFP evaluation for something in the part. As a result, the stats would never allow us to filter this part and thus can be trimmed.
+- Third, we can adjust truncation of stats over varlen columns. For example, a bound of `a<huge string>..=y<huge string>` can likely be shortened to `a..=z`. This does make the bounds less tight, which increases false positives.
+- Finally, we can trim stats for `Record`, `List`, or `Jsonb` elements or even for entire columns. Note that if we do this, we almost certainly want some way for the customer to be able to hint that a certain column is critical and should never have trimmed stats (possibly even something as simple as a LaunchDarkly flag).
+
+
+Note that it is expected that the common case will be that stats do not need trimming. This is more a safety measure.
+
+## Filtering parts
+[filtering parts]: #filtering-parts
+
+Thanks to our part statistics, we know the range of possible values any particular column may have.
+To decide whether we can filter a part or not, though, we need to know the range of possible values of _the output of the MFP's filter:_ if we can prove that the filter will never return `true` or an error value, it's safe to skip the entire part.
+
+Using the range of possible input values to compute the range of possible output values is a well-known static analysis problem, and it's commonly solved via [abstract interpretation](https://en.wikipedia.org/wiki/Abstract_interpretation). In practice, this involves:
+- Choosing an "abstract set" to approximate your data type. For part filters, we can use roughly:
+    ```rust
+    /// Approximating a range of possible non-null datums.
+    enum Values<'a> {
+      Empty, // No non-null datums.
+      Within(Datum<'a>, Datum<'a>), // An inclusive range.
+      All, // Includes all datums.
+    }
+
+    /// Approximating a set of possible `Result<Datum, Err>`;
+    /// ie. the result of `MirScalarExpr::eval`.
+    pub struct ResultSpec<'a> {
+      nullable: bool, // Could this result be null?
+      fallible: bool, // Could this result be an error?
+      values: Values<'a>, // What range of non-null values may be included?
```

Comment:
I'm slightly confused by the information carried by `Values` and `ResultSpec`. Why do we need `values: Values<'a>` in the latter instead of a single `Values<'a>`?

Reply:
Sorry, I managed to answer this myself after reviewing #19173 last week.